bigquery-storage/api/BigQueryStorage.Samples/Quickstart.cs

```diff
@@ -0,0 +1,85 @@
+﻿/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START bigquerystorage_quickstart]
+
+using Avro;
+using Avro.Generic;
+using Avro.IO;
+using Google.Cloud.BigQuery.Storage.V1;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using static Google.Cloud.BigQuery.Storage.V1.ReadSession.Types;
+
+public class QuickstartSample
+{
+    public struct Data
+    {
+        public string Name { get; set; }
+        public string State { get; set; }
+        public long Number { get; set; }
+    }
+
+    public async Task<List<Data>> QuickstartAsync(string projectId)
+    {
+        var bigQueryReadClient = BigQueryReadClient.Create();
+
+        CreateReadSessionRequest createReadSessionRequest = new CreateReadSessionRequest
+        {
+            Parent = $"projects/{projectId}",
+            ReadSession = new ReadSession
+            {
+                // This example uses baby name data from the public datasets.
+                Table = "projects/bigquery-public-data/datasets/usa_names/tables/usa_1910_current",
+                DataFormat = DataFormat.Avro,
+                ReadOptions = new TableReadOptions
+                {
+                    // Specify the columns to be projected by adding them to the selected fields.
+                    SelectedFields = { "name", "number", "state" },
+                    RowRestriction = "state = \"WA\""
+                }
+            },
+            MaxStreamCount = 1,
+        };
+        var readSession = bigQueryReadClient.CreateReadSession(createReadSessionRequest);
+
+        // Using the first stream to perform reading.
+        var readRowsStream = bigQueryReadClient.ReadRows(readSession.Streams.First().Name, 0).GetResponseStream();
+        var schema = Schema.Parse(readSession.AvroSchema.Schema);
+        var reader = new GenericDatumReader<GenericRecord>(schema, schema);
+        var dataList = new List<Data>();
+
+        await foreach (var readRowResponse in readRowsStream)
+        {
+            var byteArray = readRowResponse.AvroRows.SerializedBinaryRows.ToByteArray();
+            var decoder = new BinaryDecoder(new MemoryStream(byteArray));
+            for (int counter = 0; counter < readRowResponse.RowCount; counter++)
+            {
+                var record = reader.Read(null, decoder);
```

Comment:
I think this read method is generic, so you can do something like
```suggestion
                Data record = reader.Read<Data>(null, decoder);
```

Reply:
I don't think we can do this as we are using the GenericDataReader that spits data in form of a GenericRecord.

Reply:
It's reading to a GenericRecord because you hace the following line of code on line 67:

```
var reader = new GenericDatumReader<GenericRecord>(schema, schema);
```

You probably can do instead

```
var reader = new GenericDatumReader<BabyNamesData>(schema, schema);
```

and the you don't even have to convert field by field, the record you get will probably be what you want, like I'm suggesting.

https://avro.apache.org/docs/current/api/csharp/html/classAvro_1_1Generic_1_1GenericDatumReader.html

Reply:
Getting the below error after using it ```Unable to cast object of type 'Avro.Generic.GenericRecord' to type 'BabyNamesData'.``` I think may be we need to try with other kind of reader.

Reply:
Yes please, look into that. It'll be a little awkwad if there's a cleaner way to do this and on the sample where we are showing how to work with Avro, we do it the hard way. Thanks.

Obviously just invest a sensible amount of time into it.